File: binary_tree_preorder_traversal.java
/*
Binary Tree Preorder Traversal Total Accepted: 27626 Total Submissions: 77757 My Submissions
Given a binary tree, return the preorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [1,2,3].

Note: Recursive solution is trivial, could you do it iteratively?
*/
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>();
        if (root == null) {
            return result;
        }
        
        Stack<TreeNode> s = new Stack<TreeNode>();
        s.push(root);
        
        while(!s.isEmpty()) {
            TreeNode current = s.pop();
            result.add(current.val);
            if (current.right != null) {
                s.push(current.right);
            }
            if (current.left != null) {
                s.push(current.left);
            }
        }
        
        return result;
    }
}

File: Binary Tree Preorder Traversal.cpp



/**
    Binary Tree Preorder Traversal 
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 //first method is recursion
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        vector<int> result;
        if(root==NULL){
            return result;
        }
        preorderHelper(root,result);
        return result;
    }
    void preorderHelper(TreeNode* root, vector<int> &result){
        if(root==NULL){
            return;
        }
        result.push_back(root->val);
        if(root->left!=NULL){
            preorderHelper(root->left,result);
        }
        if(root->right!=NULL){
            preorderHelper(root->right,result);
        }
        
    }
};

//second method is iteration

/*Notice: it is very classic method. We need a stack and the method needs the following steps
(1) Push the root node into the stack.
(2) while the stack is not empty, do:
       a. pop the top node and print it.
       b. push the right child of the top node (if exists)
       c. push the left child of the top node (if exists)

*/
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        vector<int> result;
        if(root==NULL){
            return result;
        }
        stack<TreeNode*> Stack;
        //TreeNode cur = root;
        Stack.push(root);
        while(!Stack.empty()){
          TreeNode* cur = Stack.top();
          Stack.pop();
          result.push_back(cur->val);
          
          if(cur->right)
            Stack.push(cur->right);
          if(cur->left)
            Stack.push(cur->left);
            
            
        }
        
        return result;
    }
    
};